### ISSUE TITLE
hcsshim::ImportLayer failed in Win32: The system cannot find the path specified. (0x3)

### ISSUE THREAD

**Issue by dhirschfeld** [moby/moby#32838]
When trying to build a docker container for a Python data-science environment I'm getting the mentioned error right at the end of the build. Unfortunately it doesn't appear to happen on less complicated builds.

Error message at end of `docker build`
```
re-exec error: exit status 1: output: time="2017-04-26T15:58:20+10:00" level=error msg="hcsshim::ImportLayer failed in Win32: The system cannot find the path specified. (0x3) layerId=\\\\?\\C:\\ProgramData\\docker\\windowsfilter\\846463084a548600306c3f05fd185b7b46f04695bae40908f6485d3989b5e291 flavour=1 folder=C:\\Windows\\TEMP\\hcs235647079"
hcsshim::ImportLayer failed in Win32: The system cannot find the path specified. (0x3) layerId=\\?\C:\ProgramData\docker\windowsfilter\846463084a548600306c3f05fd185b7b46f04695bae40908f6485d3989b5e291 flavour=1 folder=C:\Windows\TEMP\hcs235647079
```

**Output of `docker version`:**
```
Client:
 Version:      17.06.0-dev
 API version:  1.30
 Go version:   go1.7.5
 Git commit:   d4fb626
 Built:        Wed Apr 26 03:46:47 2017
 OS/Arch:      windows/amd64

Server:
 Version:      17.06.0-dev
 API version:  1.30 (minimum version 1.24)
 Go version:   go1.7.5
 Git commit:   d4fb626
 Built:        Wed Apr 26 03:46:47 2017
 OS/Arch:      windows/amd64
 Experimental: true
```

**Output of `docker info`:**
```
Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 13
Server Version: 17.06.0-dev
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: l2bridge l2tunnel nat null overlay transparent
Swarm: inactive
Default Isolation: process
Kernel Version: 10.0 14393 (14393.1066.amd64fre.rs1_release_sec.170327-1835)
Operating System: Windows Server 2016 Standard
OSType: windows
Architecture: x86_64
CPUs: 4
Total Memory: 16GiB
Name: *******
ID: <redacted>
Docker Root Dir: C:\ProgramData\docker
Debug Mode (client): false
Debug Mode (server): false
Http Proxy: <redacted>
Https Proxy: <redacted>
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details:**
This is on a Windows Server 2016 VM running on vSphere
![image](https://cloud.githubusercontent.com/assets/881019/25420451/74137b3a-2a9b-11e7-9438-6e5b0a4d951f.png)
![image](https://cloud.githubusercontent.com/assets/881019/25420510/bb38ed60-2a9b-11e7-911c-8506c65d074b.png)

**Comment by dhirschfeld** [comment]
The Dockerfile for the build was:
```dockerfile
FROM dhirschfeld/nano-miniconda

SHELL ["powershell", "-Command", "$ErrorActionPreference='Stop'; $ProgressPreference='SilentlyContinue';"]

ADD testenv.yml C:/

RUN Start-Process conda -ArgumentList 'config','--set','channel_priority','false' -NoNewWindow -Wait; \
    Start-Process conda -ArgumentList 'config','--append','channels','conda-forge' -NoNewWindow -Wait; \
    Start-Process conda -ArgumentList 'update','--yes','--all'  -NoNewWindow -Wait; \
    Start-Process conda -ArgumentList 'env','update','--file','C:/testenv.yml'  -NoNewWindow -Wait; \
    Start-Process conda -ArgumentList 'clean','-tips' -NoNewWindow -Wait; \
    Remove-Item -Force C:/testenv.yml;

CMD ["cmd.exe"]
```

The miniconda container is available from my DockerHub account: 
https://hub.docker.com/r/dhirschfeld/nano-miniconda/

And the required `testenv.yml` is available at:
https://gist.github.com/dhirschfeld/9ef0398fa8ba757e51dc0ef3b1c4fe37

**Comment by dhirschfeld** [comment]
Just a note to mention that if I run the `nano-miniconda` container then install the Python `testenv` manually everything seems to work fine but I get the same `hcsshim::ImportLayer` error if I try to commit the container.

**Comment by dhirschfeld** [comment]
I can also confirm that I'm seeing the same error on a new Azure VM so at least it should be reproducible.

**Comment by dhirschfeld** [comment]
...and as a further data point, it also fails with the same error message on Win10 Pro Version 1703 build 1506.250.

**Comment by thaJeztah** [comment]
ping @friism @johnstep PTAL

**Comment by thaJeztah** [comment]
(is this still an issue for you @dhirschfeld ?)

**Comment by dhirschfeld** [comment]
Yes. I worked around it but then it later keeps cropping up. I update the container then can't commit it.

**Comment by dhirschfeld** [comment]
Unfortunately under normal circumstances building a container  Just Works. It's only when trying to build a very large container for Python analytics that this comes up. As noted above though the error *is* reproducible even if triggering it can be a somewhat involved process.

**Comment by dhirschfeld** [comment]
Here's another build that fails for me:
```dockerfile
FROM microsoft/windowsservercore as msbuild
SHELL ["powershell", "-Command", "$ErrorActionPreference='Stop'; $ProgressPreference='SilentlyContinue';"]
ADD buildtools C:/buildtools
WORKDIR C:/buildtools
RUN Start-Process vs_buildtools.exe -ArgumentList '--quiet','--norestart','--locale en-US','--add Microsoft.VisualStudio.Workload.VCTools','--includeRecommended','--wait' -NoNewWindow -Wait;

FROM microsoft/windowsservercore
COPY --from=msbuild 'C:\Program Files (x86)\Microsoft Visual Studio' 'C:\Program Files (x86)\Microsoft Visual Studio'
RUN setx /M PATH $($Env:PATH + ';' + 'C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\15.0\Bin')
```
```
Sending build context to Docker daemon  4.511GB
Step 1/8 : FROM microsoft/windowsservercore as msbuild
 ---> 015cd665fbdd
Step 2/8 : SHELL powershell -Command $ErrorActionPreference='Stop'; $ProgressPreference='SilentlyContinue';
 ---> Using cache
 ---> 168179012499
Step 3/8 : ADD buildtools C:/buildtools
re-exec error: exit status 1: output: time="2017-06-20T11:37:38+10:00" level=error msg="hcsshim::ImportLayer failed in Win32: The system cannot find the path specified. (0x3) layerId=\\\\?\\C:\\ProgramData\\docker\\windowsfilter\\b41d28c95f98368b73fc192cb9205700e21
6691495c1f9ac79b9b04ec4923ea2 flavour=1 folder=C:\\Windows\\TEMP\\hcs232661915"
hcsshim::ImportLayer failed in Win32: The system cannot find the path specified. (0x3) layerId=\\?\C:\ProgramData\docker\windowsfilter\b41d28c95f98368b73fc192cb9205700e216691495c1f9ac79b9b04ec4923ea2 flavour=1 folder=C:\Windows\TEMP\hcs232661915
```
The `buildtools` folder was created from the [Build Tools for Visual Studio 2017](https://www.visualstudio.com/thank-you-downloading-visual-studio/?sku=BuildTools&rel=15) installer with the command:
```
vs_buildtools.exe --locale en-US --add Microsoft.VisualStudio.Workload.VCTools --includeRecommended --layout ./buildtools
```
```
C:\dev\build\docker\buildslave> docker version
Client:
 Version:      17.06.0-dev
 API version:  1.31
 Go version:   go1.8.3
 Git commit:   5dd3073
 Built:        Sun Jun 18 23:43:19 2017
 OS/Arch:      windows/amd64

Server:
 Version:      17.06.0-dev
 API version:  1.31 (minimum version 1.24)
 Go version:   go1.8.3
 Git commit:   4f25969
 Built:        Sun Jun 18 23:47:02 2017
 OS/Arch:      windows/amd64
```

**Comment by dhirschfeld** [comment]
...so I'm finding that on anything other than simple builds the `hcsshim::ImportLayer` error happens all the time :(

It would be great to know if anyone else is also seeing it or if it's something specific to my setup. If the latter there's some hope I can work around it.

Since I've also seen the error on my home laptop running Win10 with no proxy or other corporate issues to muddy the waters, I suspect it's an actual bug  in the docker implementation.

**Comment by dtma007** [comment]
I am also seeing the same type of error message that you are seeing.
"hcsshim::ImportLayer failed in Win32: The system cannot find the path specified. (0x3) layerId=\\\\?\\C:\\ProgramData\\docker\\windowsfilter\\8965301e9e9735141b479ed3cc873107ef16c4bfbb3dc0e54d0b4827d034722a flavour=1 folder=C:\\windows\\TEMP\\hcs510325079"

In my case I'm starting off from a "microsoft/mssql-server-windows-express" as my base Docker container and installing some additional software in the container.

I have tried various things such as:
 * stopping and removing any other containers on the host machine
 * rebooting
 * retrying several times

But nothing has worked and I'm getting pretty frustrated.  I still see a "This operation returned because the timeout period expired" from time to time, but I'm able to get past that error by just rerunning a few times.  Nothing has worked so far for this "path not found" errror .

**Comment by johnstep** [comment]
@dhirschfeld I am going to try the Visual Studio scenario now.

**Comment by dhirschfeld** [comment]
Thanks @johnstep - I'm very interested to see if you can reproduce the error. Let me know if you need any more info from my side...

**Comment by johnstep** [comment]
@dhirschfeld In the Visual Studio example, have you ever made it past step 3? That step works for me, but, when I was low on free space, I ran into a similar issue. How much free space do you have on the Docker volume before starting the build?

After encountering the issue, use the following PowerShell command to get a list of the sandbox VHDs currently mounted (curious to see if these are remaining mounted):

```
(Get-Disk |? { $_.FriendlyName -eq 'Msft Virtual Disk' } |% { Get-VHD $_.Number }).Path
```

I ran into a different issue in step 7, where the COPY instruction fails due to spaces in the directory names.

**Comment by dhirschfeld** [comment]
I think I was pushing the free space on my hard-drive. On another occasion a build did error out with a message about not enough free space so I've moved docker to a new hard-drive with plenty of space.

If the error *is* due to free space issues a more relevant error message would be very helpful! On my  Win10 PC though I'm sure I had 100GB free so it *should* have worked there if it was a space issue.

I'll see if I can repro and report back the result of the powershell command.

I zipped the buildtools folder and managed to build the image though I can't remember if that was before or after I moved docker to a new hard drive. In any case I also saw the issue with the spaces which I worked around using the JSON syntax (with forward slashes).

**Comment by vikalyan** [comment]
Ran into this issue with Docker EE for Windows Server. 
https://github.com/docker/for-win/issues/905

**Comment by blues4ugrl** [comment]
@dtma007 - I am getting similar errors to you, though we are likely doing different things. I have been able to reproduce the issue in my scenario - so I'm hoping that this helps.

Running a Jenkins job on a Windows Server 2016 Datacenter edition Amazon EC2 instance. The job runs a Powershell script that does a docker pull of a Windows image from Account A, a docker tag, and then logs in to Account B and does a docker push. That part works fine. The second part of the script does a docker rmi --force to remove the copy of the image from Docker on the Windows Jenkins host. Even that actually removes the local image, but then the next time I try to run the same process, I get the errors. After hours of trial and error, I was able to get past this and reproduce and fix again twice. Below is the lengthy detail:

First error from event log on Windows Jenkins node:
hcsshim::DestroyLayer failed in Win32: Access is denied. (0x5) id=c2090764b798c63438ee74f886235307b969db55eef01c9fae1558d1fe5907ea flavour=1

Second error:
Error releasing layer sha256:18051c4005b6b29f69cdc8c8afadcedbfcf06de4604f567565dd780e63e9bbc1: hcsshim::DestroyLayer failed in Win32: Access is denied. (0x5) id=c2090764b798c63438ee74f886235307b969db55eef01c9fae1558d1fe5907ea flavour=1

Third error:
Error trying v2 registry: failed to register layer: re-exec error: exit status 1: output: time="2017-07-11T00:36:42Z" level=error msg="hcsshim::ImportLayer failed in Win32: The system cannot find the file specified. (0x2) layerId=\\\\?\\C:\\ProgramData\\docker\\windowsfilter\\a40dedbfc02b21485a4ac6aa085d196be59dccc68d57c410b568bc4e44d6ecdd flavour=1 folder=C:\\Windows\\TEMP\\hcs040092411" 
hcsshim::ImportLayer failed in Win32: The system cannot find the file specified. (0x2) layerId=\\?\C:\ProgramData\docker\windowsfilter\a40dedbfc02b21485a4ac6aa085d196be59dccc68d57c410b568bc4e44d6ecdd flavour=1 folder=C:\Windows\TEMP\hcs040092411

Fourth error:
Attempting next endpoint for pull after error: failed to register layer: re-exec error: exit status 1: output: time="2017-07-11T00:36:42Z" level=error msg="hcsshim::ImportLayer failed in Win32: The system cannot find the file specified. (0x2) layerId=\\\\?\\C:\\ProgramData\\docker\\windowsfilter\\a40dedbfc02b21485a4ac6aa085d196be59dccc68d57c410b568bc4e44d6ecdd flavour=1 folder=C:\\Windows\\TEMP\\hcs040092411" 
hcsshim::ImportLayer failed in Win32: The system cannot find the file specified. (0x2) layerId=\\?\C:\ProgramData\docker\windowsfilter\a40dedbfc02b21485a4ac6aa085d196be59dccc68d57c410b568bc4e44d6ecdd flavour=1 folder=C:\Windows\TEMP\hcs040092411

I then RENAMED the following folder:  C:\\ProgramData\\docker\\windowsfilter\\c2090764b798c63438ee74f886235307b969db55eef01c9fae1558d1fe5907ea

Ran the Powershell script again

Next Information event:
Layer sha256:18051c4005b6b29f69cdc8c8afadcedbfcf06de4604f567565dd780e63e9bbc1 cleaned up

Fifth error:
Error trying v2 registry: failed to register layer: Cannot create layer with missing parent c2090764b798c63438ee74f886235307b969db55eef01c9fae1558d1fe5907ea: GetFileAttributesEx C:\ProgramData\docker\windowsfilter\c2090764b798c63438ee74f886235307b969db55eef01c9fae1558d1fe5907ea: The system cannot find the file specified.

Sixth error:
Attempting next endpoint for pull after error: failed to register layer: Cannot create layer with missing parent c2090764b798c63438ee74f886235307b969db55eef01c9fae1558d1fe5907ea: GetFileAttributesEx C:\ProgramData\docker\windowsfilter\c2090764b798c63438ee74f886235307b969db55eef01c9fae1558d1fe5907ea: The system cannot find the file specified.

Seventh error:
Handler for POST /v1.25/images/239641032376.dkr.ecr.us-east-1.amazonaws.com/omnious/windows-container-launcher:0.1.13/tag returned error: no such id: 239641032376.dkr.ecr.us-east-1.amazonaws.com/omnious/windows-container-launcher:0.1.13

Renamed folder back and ran job again.....and it was successful!

For now, we have removed the cleanup as this is what appears to have directly caused the issue (docker rmi <image> --force). So somewhere in that command, something isn't getting cleaned up right. Renaming the directory, re-running, then naming it back seems to trigger a good cleanup. I'm still working on troubleshooting this further, and I'm still a newb, but I wanted to comment here since i'm seeing the same errors and have found a way to trigger & resolve them.

Docker version:
Client:
 Version:      1.12.2-cs2-ws-beta
 API version:  1.25
 Go version:   go1.7.1
 Git commit:   050b611
 Built:        Tue Oct 11 02:35:40 2016
 OS/Arch:      windows/amd64

Server:
 Version:      1.12.2-cs2-ws-beta
 API version:  1.25
 Go version:   go1.7.1
 Git commit:   050b611
 Built:        Tue Oct 11 02:35:40 2016
 OS/Arch:      windows/amd64

Docker info:
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 2
Server Version: 1.12.2-cs2-ws-beta
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: nat null overlay
Swarm: inactive
Default Isolation: process
Kernel Version: 10.0 14393 (14393.1198.amd64fre.rs1_release_sec.170427-1353)
Operating System: Windows Server 2016 Datacenter
OSType: windows
Architecture: x86_64
CPUs: 2
Total Memory: 8 GiB
Name: EC2AMAZ-5AK76IV
ID: VZIV:QABT:63VB:KUOW:4FOR:BTTB:QULZ:U7KS:CIZK:QWTU:63SI:65EG
Docker Root Dir: C:\ProgramData\docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

**Comment by friism** [comment]
Possibly related: https://github.com/moby/moby/issues/31253#issuecomment-314525571

**Comment by mback2k** [comment]
I am experiencing the same issue randomly on Windows 10 1709 using a Windows Server Standard Core Container 1709 with HyperV isolation. It seems like higher numbers of files changed in the CoW layer triggering the issue more often. Since I install and update CI tools inside the Dockerfile I cannot reduce the number of changed files easily. I just uploaded some diagnostic data with the ID `6B3344D2-36A6-44A4-89BB-764B62C76472/2017-11-10_23-31-44`. Can you take a look at it, please?

**Comment by mback2k** [comment]
The following Dockerfile can be used to re-produce this issue most of the time on the following Docker version on Windows 10 Pro 1709 (Build 16299.19):
```
Client:
 Version:      17.09.0-ce
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:40:09 2017
 OS/Arch:      windows/amd64

Server:
 Version:      17.09.0-ce
 API version:  1.32 (minimum version 1.24)
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:50:27 2017
 OS/Arch:      windows/amd64
 Experimental: true
```

Dockerfile:
```
# escape=`

FROM microsoft/windowsservercore:1709_KB4043961

SHELL ["powershell", "-command"]

RUN Invoke-WebRequest "http://www.7-zip.org/a/7z1604-x64.exe" -OutFile 7z1604-x64.exe;
RUN Start-Process -FilePath "C:\7z1604-x64.exe" -ArgumentList /S -NoNewWindow -PassThru -Wait;
RUN Remove-Item 7z1604-x64.exe;

RUN Invoke-WebRequest "http://repo.msys2.org/distrib/msys2-x86_64-latest.tar.xz" -OutFile msys2-x86_64-latest.tar.xz;
RUN Start-Process -FilePath "C:\Program` Files\7-Zip\7z.exe" -ArgumentList e, msys2-x86_64-latest.tar.xz -NoNewWindow -PassThru -Wait;
RUN Start-Process -FilePath "C:\Program` Files\7-Zip\7z.exe" -ArgumentList x, msys2-x86_64-latest.tar, `-oC:\ -NoNewWindow -PassThru -Wait;
RUN Remove-Item msys2-x86_64-latest.tar.xz; Remove-Item msys2-x86_64-latest.tar;

RUN Write-Host 'Updating MSYSTEM and MSYSCON ...'; `
	[Environment]::SetEnvironmentVariable('MSYSTEM', 'MSYS2', [EnvironmentVariableTarget]::Machine); `
	[Environment]::SetEnvironmentVariable('MSYSCON', 'defterm', [EnvironmentVariableTarget]::Machine);

# For some reason bash.exe has to be called first twice since we are not building interactive.
RUN C:\msys64\usr\bin\bash.exe --login -c 'exit 0'; `
	C:\msys64\usr\bin\bash.exe --login -c 'exit 0'; `
	C:\msys64\usr\bin\bash.exe --login -c 'pacman -Syu --noconfirm';

RUN C:\msys64\usr\bin\bash.exe --login -c 'exit 0'; `
	C:\msys64\usr\bin\bash.exe --login -c 'exit 0'; `
	C:\msys64\usr\bin\bash.exe --login -c 'pacman -Su  --noconfirm';
```

The error happens right after completion of the 2nd to last RUN statement. New diagnostic ID `6B3344D2-36A6-44A4-89BB-764B62C76472/2017-11-11_00-03-45`.

**Comment by mback2k** [comment]
It can also be re-produced with the following Docker version:
```
Client:
 Version:      17.10.0-ee-preview-3
 API version:  1.33
 Go version:   go1.8.4
 Git commit:   1649af8
 Built:        Fri Oct  6 17:52:28 2017
 OS/Arch:      windows/amd64

Server:
 Version:      17.10.0-ee-preview-3
 API version:  1.34 (minimum version 1.24)
 Go version:   go1.8.4
 Git commit:   b8571fd
 Built:        Fri Oct  6 18:01:48 2017
 OS/Arch:      windows/amd64
 Experimental: true
```

**Comment by thepill** [comment]
For me, a similar error happens when i pulling a image on a server 2016 from a registry where the images has been pushed from windows 10 1709:

```
failed to register layer: re-exec error: exit status 1: output: time="2017-11-27T09:31:11+01:00" level=error msg="hcsshim::ImportLayer failed in Win32: The system cannot find the file specified. (0x2) layerId=\\\\?\\C:\\ProgramData\\docker\\windowsfilter\\8f9191263fd2a5ae93c29ba554426c4adce8fcf333c1b9bf4f217556650d628e flavour=1 folder=C:\\Windows\\TEMP\\hcs456696507"
hcsshim::ImportLayer failed in Win32: The system cannot find the file specified. (0x2) layerId=\\?\C:\ProgramData\docker\windowsfilter\8f9191263fd2a5ae93c29ba554426c4adce8fcf333c1b9bf4f217556650d628e flavour=1 folder=C:\Windows\TEMP\hcs456696507
```

If i build the exactly image directly on the server 2016 node, it will work.
I tried `17.11.0-ce` and `17.09.0-ce` on my windows 10 node-> same result.
`17.06.2-ee-5` is running on my windows server 2016 node.

**Comment by archisgore** [comment]
I can't build a simple hello world based on the instructions I found in: https://hub.docker.com/r/microsoft/iis/

My contents contains literally an 80-byte index.html that has the text "Hello World", and I'm building using VSTS Online.

```
FROM microsoft/iis:windowsservercore-1709

RUN mkdir C:\site

RUN powershell -NoProfile -Command \
    Import-module IISAdministration; \
    New-IISSite -Name "Site" -PhysicalPath C:\site -BindingInformation "*:80:"

EXPOSE 80

ADD content/ /site
```

**Comment by dhirschfeld** [comment]
@archisgore - please open a new issue for problems unrelated to the original topic.

**Comment by archisgore** [comment]
Oh I meant I get the same error. Not a new topic. I just didn't want to paste the redundant error info. I pasted the Dockerfile mainly to illustrate that I am doing nothing fancy/magical/complex. Sorry about that.

**Comment by dhirschfeld** [comment]
Out of interest, how much free hard drive space do you have?

**Comment by archisgore** [comment]
I don't know honestly. I used VSTS Online's "VSTS Agent" or whatever they call it. I deleted my account out of frustration. This build works now that I booked an Azure instance though with 30 gigs free though.

**Comment by vikalyan** [comment]
@dhirschfeld - This happens no matter the size of the disk. I have an Azure VM with 2TiB and 64 GiB of RAM. Still run into this problem.

**Comment by dhirschfeld** [comment]
I'm on a new server myself with the latest binaries from https://master.dockerproject.org/ and I haven't been able to reproduce the error (yet) with some very large (15GB) complicated builds.

Might be worthwhile seeing if it's reproducible with the latest binaries.
```
λ docker version
<snip>
Server:
 Version:      master-dockerproject-2017-12-01
 API version:  1.35 (minimum version 1.24)
 Go version:   go1.9.2
 Git commit:   ac68411
 Built:        Fri Dec  1 23:52:43 2017
 OS/Arch:      windows/amd64
 Experimental: false
```

**Comment by atrauzzi** [comment]
On Windows 10 Pro, trying to get a build agent mastered locally, using chocolatey to install lots of Visual Studio guts.  If you want something that thrashes the filesystem, I highly recommend it!

Got this error, similar to what's already been reported here:

```
re-exec error: exit status 1: output: time="2017-12-13T14:02:28-06:00" level=error msg="hcsshim::ImportLayer failed in Win32: The system
cannot find the path specified. (0x3) layerId=\\?\C:\ProgramData\Docker\windowsfilter\34175b12d23c6ed1abc7827e810ce0ba5cb5e0727a769b96c69
482cd9b6cc744 flavour=1 folder=C:\WINDOWS\TEMP\hcs508432187"
hcsshim::ImportLayer failed in Win32: The system cannot find the path specified. (0x3) layerId=\\?\C:\ProgramData\Docker\windowsfilter\34
175b12d23c6ed1abc7827e810ce0ba5cb5e0727a769b96c69482cd9b6cc744 flavour=1 folder=C:\WINDOWS\TEMP\hcs508432187
```

```
Docker version 17.09.1-ce, build 19e2cf6
```

Lots of free disk space.  If I retry the build, it picks up from the previous step and I can eventually get it through.  Unfortunately it's a huge pain because each step involves downloading a large amount of data and then writing it to the disk.

**Comment by vikalyan** [comment]
@atrauzzi  - Yes this bug is very annoying. I have found that retrying is the only way to go; perhaps ensure that you are not using --force-rm and --no-cache so that you can cache and reuse the intermediate layers?

**Comment by vikalyan** [comment]
@dhirschfeld  - Expecting everyone to upgrade to 1709 (which is a server core installation) is not a reasonable ask. The stability fixes if any should be rolled out to folks on 1607 as well.

**Comment by dhirschfeld** [comment]
My comment was for informational purposes to help out the person debugging this.

I agree docker should work on all the latest Windows versions.

To further help out the person debugging this you could update to the latest docker master binaries and see if the problem persists.

I'm no longer experiencing the problem but I changed (at least) two things. To figure out which of those is relevant (server version or docker version) others can fairly easily test at least the docker version...

**Comment by jhowardmsft** [comment]
@dhirschfeld Can you provide a new link to your testenv.yml file? The link above gives a 404

**Comment by dhirschfeld** [comment]
@jhowardmsft - https://gist.github.com/dhirschfeld/03887a81080345997a70230b60941fe3

**Comment by mback2k** [comment]
I am still running into this issue with the Dockerfile posted above and the following versions:
```Client:
 Version:       17.12.0-ce-rc3
 API version:   1.35
 Go version:    go1.9.2
 Git commit:    80c8033
 Built: Thu Dec 14 00:39:41 2017
 OS/Arch:       windows/amd64

Server:
 Engine:
  Version:      17.12.0-ce-rc3
  API version:  1.35 (minimum version 1.24)
  Go version:   go1.9.2
  Git commit:   80c8033
  Built:        Thu Dec 14 00:49:52 2017
  OS/Arch:      windows/amd64
  Experimental: true```

**Comment by danechambers** [comment]
I hate to pile on here, but I am running into this error as well. I would like to contribute (yet) another dockerfile that I am able to reliably(ish) reproduce the issue on:
[Dockerfile.txt](https://github.com/moby/moby/files/1569732/Dockerfile.txt)

>dane@work:test> docker info
>Containers: 5
> Running: 1
> Paused: 0
> Stopped: 4
>Images: 42
>Server Version: 17.09.1-ce
>Storage Driver: windowsfilter
> Windows:
>Logging Driver: json-file
>Plugins:
> Volume: local
> Network: ics l2bridge l2tunnel nat null overlay transparent
> Log: awslogs etwlogs fluentd json-file logentries splunk syslog
>Swarm: inactive
>Default Isolation: hyperv
>Kernel Version: 10.0 16299 (16299.15.amd64fre.rs3_release.170928-1534)
>Operating System: Windows 10 Pro
>OSType: windows
>Architecture: x86_64
>CPUs: 8
>Total Memory: 15.91GiB
>Docker Root Dir: C:\ProgramData\Docker
>Debug Mode (client): false
>Debug Mode (server): true
> File Descriptors: -1
> Goroutines: 35
> System Time: 2017-12-18T15:08:32.7522212-08:00
> EventsListeners: 0
>Registry: https://index.docker.io/v1/
>Experimental: true
>Insecure Registries:
> 127.0.0.0/8
>Live Restore Enabled: false

**Comment by jhowardmsft** [comment]
OK, I can repro with the dockerfile from @mback2k above, and with a cut-down smaller repro.

In that dockerfile, some files are being left in the recycle bin which have "odd" filenames. It looks currently like a golang bug being able to interpret Unicode characters in filenames. I just used these files, and wrote a small program which does ioutil.ReadDir to enumerate through files, and call os.Stat on each file. Stat calls GetFileAttributeEx which fails in the same way as debugging docker.

**Comment by jhowardmsft** [comment]
@danechambers How "reliably(ish)"? Using your dockerfile and moby/moby master branch on Windows 1709/fall creators update/build 16299, I've done two builds (using Hyper-V isolation) successfully.

**Comment by mback2k** [comment]
@mback2k Thanks a lot for your help. Can you please post here once the fix makes it into a new Docker release?

**Comment by jhowardmsft** [comment]
@mback2k Once https://github.com/moby/moby/pull/35924 is merged, an updated binary will be available shortly after at https://master.dockerproject.org

**Comment by jhowardmsft** [comment]
@dhirschfeld Can you confirm if the very original scenario (the one using `testenv.yml`) is still failing doing a docker build? You mentioned things about updating the container and committing. If still an issue, can you provide full repro-steps? I couldn't repro docker build failing using testenv.yml you posted in December (https://github.com/moby/moby/issues/32838#issuecomment-351864793), and the dockerfile at the top (https://github.com/moby/moby/issues/32838#issuecomment-297252492). My repro environment is using Windows Server build 1709/Fall Creators Update/build 16299, and docker built using moby/moby master as of this morning (which doesn't have the HCSShim fix linked to above). I had to build with Hyper-V isolation as the base image is from an older version of Windows - ie `docker build --isolation=hyperv .`.

**Comment by jhowardmsft** [comment]
@dtma007 If you are still seeing the issue, can you open a new issue in this repro (to avoid clouding this issue with more "me-toos") and provide:
- The dockerfile
- Any supporting files required by the dockerfile
- The Windows build information
- The output of docker info and docker version
- Any additional repro steps required.

**Comment by jhowardmsft** [comment]
@blues4ugrl Similar to my comment immediately above to @dtma007, if you are hitting this can you also open a new issue with the requested information. Thanks.

**Comment by jhowardmsft** [comment]
@thepill Again, could you open a new issue with the above information if you are still hitting this.

**Comment by jhowardmsft** [comment]
@dhirschfeld For the second repro you provided, similar to @johnstep, I was able to get past step 3 using the same host described in https://github.com/moby/moby/issues/32838#issuecomment-355109858

**Comment by jhowardmsft** [comment]
@archisgore Just to confirm so others historically reading this issue, I couldn't repro your specific issue (and you stated as much above), so perhaps it was running out of disk space.

**Comment by jhowardmsft** [comment]
@atrauzzi Same comments as to various others above. Please open a new issue with the full set of repro steps and requested information if you are still hitting this issue.

**Comment by danechambers** [comment]
@jhowardmsft I ran some tests using my dockerfile. My steps were:
```
docker build -t testbuild .
```
if the build failed, then 
```
docker rm $(docker ps -aq)
docker rmi $(docker images -q -f dangling=true)
```
if the build succeeded, then 
```
docker rmi testbuild
```
Rinse repeat. 
It took me six tries to reproduce the error twice. The error happened the first time on the second try and then again on the sixth try. When the error did occur, it seemed to occur after the TCP Activation was being installed:
![20180103snag-07](https://user-images.githubusercontent.com/2801346/34547320-2b22769e-f0af-11e7-8072-8ea675721a03.jpg)
Let me know if I can provide you with more information!

**Comment by jhowardmsft** [comment]
@danechambers Can you provide your dockerfile; output of docker info and docker version; your dockerfile; supporting files needed by the dockerfile if any; Windows build information so I can try to repro and debug.

**Comment by danechambers** [comment]
@jhowardmsft 
Thank you so much for your help!
[Dockerfile.txt](https://github.com/moby/moby/files/1605396/Dockerfile.txt)
```
dane@work:system32> docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 26
Server Version: 17.09.1-ce
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: ics l2bridge l2tunnel nat null overlay transparent
 Log: awslogs etwlogs fluentd json-file logentries splunk syslog
Swarm: inactive
Default Isolation: hyperv
Kernel Version: 10.0 16299 (16299.15.amd64fre.rs3_release.170928-1534)
Operating System: Windows 10 Pro
OSType: windows
Architecture: x86_64
CPUs: 8
Total Memory: 15.91GiB
Docker Root Dir: C:\ProgramData\Docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: -1
 Goroutines: 27
 System Time: 2018-01-04T19:02:46.6463128-08:00
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```
```
dane@work:system32> docker version
Client:
 Version:      17.09.1-ce
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   19e2cf6
 Built:        Thu Dec  7 22:22:26 2017
 OS/Arch:      windows/amd64

Server:
 Version:      17.09.1-ce
 API version:  1.32 (minimum version 1.24)
 Go version:   go1.8.3
 Git commit:   19e2cf6
 Built:        Thu Dec  7 22:32:40 2017
 OS/Arch:      windows/amd64
 Experimental: true
```
```
System Information report written at: 01/04/18 19:05:59
[System Summary]

Item	Value	
OS Name	Microsoft Windows 10 Pro	
Version	10.0.16299 Build 16299	
Other OS Description 	Not Available	
OS Manufacturer	Microsoft Corporation	
System Manufacturer	Dell Inc.	
System Model	Dell Precision M3800	
System Type	x64-based PC	
System SKU	Dell Precision M3800	
Processor	Intel(R) Core(TM) i7-4712HQ CPU @ 2.30GHz, 2301 Mhz, 4 Core(s), 8 Logical Processor(s)	
BIOS Version/Date	Dell Inc. A10, 8/17/2015	
SMBIOS Version	2.7	
Embedded Controller Version	255.255	
BIOS Mode	UEFI	
BaseBoard Manufacturer	Dell Inc.	
BaseBoard Model	Not Available	
BaseBoard Name	Base Board	
Platform Role	Mobile	
Secure Boot State	On	
PCR7 Configuration	Elevation Required to View	
Windows Directory	C:\WINDOWS	
System Directory	C:\WINDOWS\system32	
Boot Device	\Device\HarddiskVolume1	
Locale	United States	
Hardware Abstraction Layer	Version = "10.0.16299.15"	
User Name	<user domain redacted/>	
Time Zone	Pacific Standard Time	
Installed Physical Memory (RAM)	16.0 GB	
Total Physical Memory	15.9 GB	
Available Physical Memory	4.07 GB	
Total Virtual Memory	22.0 GB	
Available Virtual Memory	2.47 GB	
Page File Space	6.11 GB	
Page File	C:\pagefile.sys	
Virtualization-based security	Running	
Virtualization-based security Required Security Properties		
Virtualization-based security Available Security Properties	Base Virtualization Support, Secure Boot, DMA Protection	
Virtualization-based security Services Configured		
Virtualization-based security Services Running		
Device Encryption Support	Elevation Required to View	
A hypervisor has been detected. Features required for Hyper-V will not be displayed.		
```
Full system info dump: 
[SystemInfo.txt](https://github.com/moby/moby/files/1605439/SystemInfo.txt)

**Comment by jhowardmsft** [comment]
@danechambers Thanks. Investigating. This looks like a different issue. I was able to repro on the third attempt.

**Comment by jhowardmsft** [comment]
@danechambers Just an FYI, still investigating, but this particular one is a Windows bug, external to Docker. The RS1 utility VM hosting the container is bug-checking during shutdown. It's not clear why yet, I'll be handing it off to others internally @ MS to analyze further. It's also shown a bug in docker which I will submit a fix for - the actual error above is extremely misleading as it should never have tried to commit the layer of a corrupt filesystem due to the shutdown bugcheck earlier.  Very occasionally I can also hit another failure which I'll also investigate if I can repro it again.

**Comment by steve-thibault** [comment]
This happens to me every time I try to build a visual studio build tools layer. I can't get past it. Is there a work around? What info should I post to help with the investigation?

**Comment by steve-thibault** [comment]
PS C:\Users\User> docker info
Containers: 3
 Running: 1
 Paused: 0
 Stopped: 2
Images: 8
Server Version: 17.09.1-ce
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: ics l2bridge l2tunnel nat null overlay transparent
 Log: awslogs etwlogs fluentd json-file logentries splunk syslog
Swarm: inactive
Default Isolation: hyperv
Kernel Version: 10.0 16299 (16299.15.amd64fre.rs3_release.170928-1534)
Operating System: Windows 10 Enterprise N
OSType: windows
Architecture: x86_64
CPUs: 2
Total Memory: 8GiB
Name: dockerhost-st
ID: YX3T:IAC7:VTZX:USD5:HH5I:KVRI:WH23:ZD75:AYFP:IJFA:NKOG:QUQT
Docker Root Dir: C:\ProgramData\Docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: -1
 Goroutines: 37
 System Time: 2018-01-09T07:52:37.0459779-05:00
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 dockerhub:5000
 127.0.0.0/8
Live Restore Enabled: false

Client:
 Version:      17.09.1-ce
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   19e2cf6
 Built:        Thu Dec  7 22:22:26 2017
 OS/Arch:      windows/amd64

Server:
 Version:      17.09.1-ce
 API version:  1.32 (minimum version 1.24)
 Go version:   go1.8.3
 Git commit:   19e2cf6
 Built:        Thu Dec  7 22:32:40 2017
 OS/Arch:      windows/amd64
 Experimental: false


Dockerfile: 
FROM microsoft/windowsservercore:1709
LABEL Author Steven Thibault 
LABEL Author_Email steve_thibault@waters.com

SHELL ["powershell.exe", "-ExecutionPolicy", "Bypass", "-Command"]

ENV VS_CHANNEL_URI=https://aka.ms/vs/15/release/799c44140/channel \
    VS_BUILDTOOLS_URI=https://aka.ms/vs/15/release/799c44140/vs_buildtools.exe \
    VS_BUILDTOOLSZIP_SHA256=1278BBFAE2D546EFC2AEEB250F27F720E3670C74C79D57E0447C85B007FFD147 \
    VS_BUILDTOOLS_SHA256=FA29EB83297AECADB0C4CD41E54512C953164E64EEDD9FB9D3BF9BD70C9A2D29 \
    NUGET_URI=http://rdeidgart:8081/artifactory/Tools/nuget/v4.1.0/nuget.exe \
    NUGET_SHA256=4C1DE9B026E0C4AB087302FF75240885742C0FAA62BD2554F913BBE1F6CB63A0 \
    JFROG_URI=http://rdeidgart:8081/artifactory/Tools/jfrog_cli/jfrog.exe \
    JFROG_SHA256=2F4501E29DB277D00C60DF51B9736DB6F012AC6D159F33E86632495F61E8AAB2 \
    COLLECT_SHA256=1845E87090C27A8ED5A10F447DD455B783DE4824C74A68505785BB2678D2E04C \
    NODEJS_URI=http://rdeidgart:8081/artifactory/Tools/nodejs/node-v9.3.0-x64.msie \
    NODEJS_SHA256=FE04C4A19BF6C400C1EA2F984E15F2B1E440D627681BFC97DB92E243D42E5185 \
    TOOLS_FOLDER='C:\\Tools'
    

# Install tools 
COPY "./tools/" C:/Tools/

RUN [System.Environment]::SetEnvironmentVariable('PATH', "\"${env:PATH};$env:TOOLS_FOLDER\;\"", 'Machine'); 

# Download vs_buildtools.exe
RUN $ErrorActionPreference = 'Stop'; \
    $ProgressPreference = 'SilentlyContinue'; \
    $VerbosePreference = 'Continue'; \
    Write-Host "Downloading vs_buildtools.exe..."; \
    Remove-Item alias:curl; \    
    Invoke-WebRequest -Uri $env:VS_BUILDTOOLS_URI -OutFile C:\vs_buildtools.exe; \
    curl -uicsbuildmanager:AP46B8xKsLFjNDwk -o "$env:TOOLS_FOLDER\\collect.exe"  http://rdeidgart:8081/artifactory/Tools/LogCollector/Collect.exe; \
    curl -uicsbuildmanager:AP46B8xKsLFjNDwk -o "$env:TOOLS_FOLDER\\nuget.exe" $env:NUGET_URI ; \
    curl -uicsbuildmanager:AP46B8xKsLFjNDwk -o "$env:TOOLS_FOLDER\\jfrog.exe" $env:JFROG_URI; \
    curl -uicsbuildmanager:AP46B8xKsLFjNDwk -o "$env:TOOLS_FOLDER\\nodejs-v9.3.0-x64.msi" http://rdeidgart:8081/artifactory/Tools/nodejs/node-v9.3.0-x64.msi; \
    Write-Host "Verifying download hash..."; \
    if ((Get-FileHash -Path C:\vs_buildtools.exe -Algorithm SHA256).Hash -ne $env:VS_BUILDTOOLS_SHA256) { throw 'vs Build Tools Download hash does not match' } \
    if ((Get-FileHash -Path C:\Tools\nodejs-v9.3.0-x64.msi -Algorithm SHA256).Hash -ne $env:NODEJS_SHA256) { throw 'Nodejs Download hash does not match' } \
    if ((Get-FileHash -Path C:\Tools\nuget.exe -Algorithm SHA256).Hash -ne $env:NUGET_SHA256) { throw 'NUGET Download hash does not match' } \
    if ((Get-FileHash -Path C:\Tools\collect.exe -Algorithm SHA256).Hash -ne $env:COLLECT_SHA256) { throw 'collect Download hash does not match' } \
    Write-Host "Build Tools Layer complete.";

# Install Visual Studio Build Tools
RUN $ErrorActionPreference = 'Stop'; \
    $VerbosePreference = 'Continue'; \
    Write-Host "Starting build tools install..."; \
    $p = Start-Process -Wait -PassThru -FilePath C:\vs_buildtools.exe -ArgumentList '--quiet --nocache --wait --norestart --all \
                                                                                     --add Microsoft.VisualStudio.Component.TestTools.Core \
                                                                                     --add Microsoft.VisualStudio.Component.TestTools.MicrosoftTestManager \
                                                                                     --remove Microsoft.VisualStudio.Component.Windows10SDK.10240 \
                                                                                     --remove Microsoft.VisualStudio.Component.Windows10SDK.10586 \
                                                                                     --remove Microsoft.VisualStudio.Component.Windows10SDK.14393 \
                                                                                     --remove Microsoft.VisualStudio.Component.Windows81SDK \
                                                                                     --installPath C:\BuildTools'; \
    if ($ret = $p.ExitCode) {C:\Tools\collect.exe; throw ('Install failed with exit code 0x{0:x}' -f $ret) } else { Write-Host "Install completed successfully"; } \
    Remove-Item -Force -Recurse -Path  C:\vs_buildtools.exe;

# Install Nodejs
RUN $ErrorActionPreference = 'Stop'; \
    $VerbosePreference = 'Continue'; \
    Write-Host "Starting nodejs install..."; \
    $p = Start-Process -Wait -PassThru -FilePath "$env:TOOLS_FOLDER\\nodejs-v9.3.0-x64.msi" -ArgumentList '/quiet  /norestart'; \
    if ($ret = $p.ExitCode) { collect.exe; throw ('Install failed with exit code 0x{0:x}' -f $ret) } else {'Node instaled successfully'}; 
	
# Use shell form to start developer command prompt and any other commands specified
SHELL ["cmd.exe", "/s", "/c"]
ENTRYPOINT C:\BuildTools\Common7\Tools\VsDevCmd.bat &&


# Default to PowerShell console running within developer command prompt environment
CMD ["powershell.exe", "-nologo"]

**Comment by jhowardmsft** [comment]
I've submitted https://github.com/moby/moby/pull/35967 for a fix to docker to report the error correctly: https://github.com/moby/moby/issues/32838#issuecomment-356005845. The Windows bug as for why the UtilityVM crashes or hangs and is terminated after a timeout is still being investigated.

**Comment by jhowardmsft** [comment]
@steve-thibault Best I can say at this moment is to test using a daemon with #35967 to see if you see a different error now like I state in that PR such as `encountered an error during WaitTimeout: failure in a Windows system call: The compute system exited unexpectedly. (0xc0370106)`. We are still investigate the underlying cause of that failure, but it's not a docker bug, it's a platform bug.

**Comment by steve-thibault** [comment]
@jhowardmsft Thanks for the response. I will give that a try. As an additional note, I created a new windows 10 system, exact same template as the current, and I am testing now. It seems to be working better, however, I am still building the image. If this new system works, I will report back. Possibly there is some sort of gradual buildup (yes technical term :-P ) that leads to this?

**Comment by steve-thibault** [comment]
@jhowardmsft As promised, the image did build on the new system at least past the problem layer. So a clean windows 10 1709, from the same template as the other, seems to be working so far.

**Comment by jhowardmsft** [comment]
@steve-thibault It appears that the platform bug is a race condition so will be intermittent. There is a chance if you were to rebuild a few times, you will hit the failure. (Assuming it is the same bug you've been hitting)

**Comment by gwellington** [comment]
Not sure if I should create a new issue, but we seem to be running into a similar intermittent issue?   This occurs when extracting layers via 'docker pull' which emits the following.  I've also tried mucking with ACLs under E:\Docker\ and that doesn't seem to fix anything either even if giving EVERYONE full-control.  Our EC2 node will come up fine, but will randomly get into this state where you can no longer extract images.  We haven't been able to resolve this issue without spinning up a new instance.

`failed to register layer: re-exec error: exit status 1: output: time="2018-01-11T18:16:04Z" level=error msg="hcsshim::Im
portLayer failed in Win32: Access is denied. (0x5) layerId=\\\\?\\E:\\docker\\windowsfilter\\b6888696dee4442c3d5709709bd
a7fd57192aca264e3f58ccd0fd87c1e5d228b flavour=1 folder=C:\\Users\\myhandle\\AppData\\Local\\Temp\\hcs873555083"
hcsshim::ImportLayer failed in Win32: Access is denied. (0x5) layerId=\\?\E:\docker\windowsfilter\b6888696dee4442c3d5709
709bda7fd57192aca264e3f58ccd0fd87c1e5d228b flavour=1 folder=C:\Users\myhandle\AppData\Local\Temp\hcs873555083`

Running the Debug-ContainerHost.ps1 script shows green across the board.  
`
Containers: 76
 Running: 0
 Paused: 0
 Stopped: 76
Images: 3
Server Version: 17.06.2-ee-6
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: l2bridge l2tunnel nat null overlay transparent
 Log: awslogs etwlogs fluentd json-file logentries splunk syslog
Swarm: inactive
Default Isolation: process
Kernel Version: 10.0 14393 (14393.2007.amd64fre.rs1_release.171231-1800)
Operating System: Windows Server 2016 Datacenter
OSType: windows
Architecture: x86_64
CPUs: 8
Total Memory: 32GiB
Name: Host
ID: VZIV:QABT:63VB:KUOW:4FOR:BTTB:QULZ:U7KS:CIZK:QWTU:63SI:65EG
Docker Root Dir: E:\docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false


Showing output from: docker version
Client:
 Version:      17.06.2-ee-6
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   e75fdb8
 Built:        Mon Nov 27 22:46:09 2017
 OS/Arch:      windows/amd64

Server:
 Version:      17.06.2-ee-6
 API version:  1.30 (minimum version 1.24)
 Go version:   go1.8.3
 Git commit:   e75fdb8
 Built:        Mon Nov 27 22:55:16 2017
 OS/Arch:      windows/amd64
 Experimental: false`

**Comment by jhowardmsft** [comment]
@gwellington This is not the same. Please open a new issue with repro steps (if possible) and the information I've asked for from various folks higher up this thread. I assume once the node gets into this state, it's stuck. If there's any info you can additionally provide to get into that state, would be appreciated so I can try to debug (won't be immediately though).

@darrenstahlmsft I don't recall having seen this before. Not sure if you have. FYI only.

**Comment by am11** [comment]
Getting "System cannot find the path specified" error intermittently with the following dockerfile:

<details>
  <summary><ins>Click to expand</ins></summary>

```dockerfile
FROM microsoft/windowsservercore

RUN powershell curl https://aka.ms/vs/15/release/vs_buildtools.exe -O vs_buildtools.exe
RUN powershell \
      curl https://github.com/git-for-windows/git/releases/download/v2.15.1.windows.2/Git-2.15.1.2-64-bit.exe
        -O git-setup.exe
RUN powershell curl https://cmake.org/files/v3.10/cmake-3.10.0-win64-x64.msi -O cmake-setup.msi

RUN powershell Start-Process git-setup.exe -ArgumentList '/VERYSILENT', '/SUPPRESSMSGBOXES','/CLOSEAPPLICATIONS', '/DIR=C:\git' -Wait

RUN msiexec /i cmake-setup.msi /quiet /norestart &&   \
    setx PATH "%ProgramFiles%\CMake\bin;%PATH%"

RUN mkdir c:\buildtools &&                            \
    vs_buildtools.exe                                 \
      --layout c:\buildtools                          \
      --lang en-US                                    \
      --add Microsoft.VisualStudio.Workload.VCTools   \
      --add Microsoft.Net.Component.4.6.TargetingPack \
      --add Microsoft.Net.Component.4.5.TargetingPack \
      --includeRecommended --passive --wait

RUN C:\buildtools\vs_setup.exe --noweb --passive --norestart --wait
```
</details><br/>

Error:

```
re-exec error: exit status 1: output: time="2018-01-13T16:34:17+02:00" level=error msg="hcsshim::ImportLayer failed in Win32: The system cannot find the path specified. (0x3) layerId=\\\\?\\C:\\ProgramData\\Docker\\windowsfilter\\7d4f2604e1b63a651d508636a08bead38e60a7a65512fbd67a2f5b3d4532df28 flavour=1 folder=C:\\ProgramData\\Docker\\tmp\\hcs885362571"
hcsshim::ImportLayer failed in Win32: The system cannot find the path specified. (0x3) layerId=\\?\C:\ProgramData\Docker\windowsfilter\7d4f2604e1b63a651d508636a08bead38e60a7a65512fbd67a2f5b3d4532df28 flavour=1 folder=C:\ProgramData\Docker\tmp\hcs885362571
```

Tested with:

* `Docker version 17.12.0-ce, build c97c6d6` and dev build:
* `Docker version 0.0.0-dev, build 6764ce92`

When using dev build `6764ce92`, re-attempting couple of (2-3) times succeeds the `vs_setup` step.

OS info:

```powershell
~\sources> systeminfo | findstr /B /C:"OS Name" /C:"OS Version"
OS Name:                   Microsoft Windows 10 Enterprise
OS Version:                10.0.16299 N/A Build 16299
```

**Comment by nick202** [comment]
For everyone experiencing this issue using Azure Windows Server 2016, make sure you are using the new semi-annual release channel of Windows Server 2016 which is currently "Windows Server 2016, version 1709". As per the following article, you cannot upgrade an existing Windows Server 2016 LTSC VM to this version. Many of the latest microsoft docker images require 1709 as a base dependency like the asp-net dotnet framework images. 

Side Note: The default dockerfile created by Visual Studio will use the 1709 images.

https://blogs.technet.microsoft.com/windowsserver/2017/10/26/faq-on-windows-server-version-1709-and-semi-annual-channel/
https://docs.docker.com/engine/installation/windows/ee-preview/#windows-10-fall-creators-update

**Comment by Tra5is** [comment]
I'm also getting the same error while building images. Typically using chocolatey packages.

One interesting thing I noticed today was that the containers started for the failing step in the build are missing the logging driver information. Is this normal?

Also, is the error associated with installers that use the temp folder which may or may not be cleaned up before the container's changes are committed?

```
Step 6/12 : RUN cinst microsoft-build-tools --confirm   --package-parameters "          --allWorkloads          --includeRecommended            --includeOptional"
 ---> Using cache
 ---> 9021a25ea8ac
Step 7/12 : RUN cinst dotnetcore-sdk --confirm;         cinst dotnet4.6.1 --confirm;    cinst dotnet4.6.2 --confirm;
 ---> Running in 6962984ea314
Chocolatey v0.10.8
Installing the following packages:
dotnetcore-sdk
...
re-exec error: exit status 1: output: time="2018-01-25T09:12:42-08:00" level=error msg="hcsshim::ImportLayer failed in Win32: The system cannot find the path specified. (0x3) layerId=\\\\?\\C:\\ProgramData\\Docker\\windowsfilter\\6b843fbbed7dfc160dbddbaa40228e095d3cc95260b3aab694ab06f179f83bcc flavour=1 folder=C:\\ProgramData\\Docker\\tmp\\hcs925889499"
hcsshim::ImportLayer failed in Win32: The system cannot find the path specified. (0x3) layerId=\\?\C:\ProgramData\Docker\windowsfilter\6b843fbbed7dfc160dbddbaa40228e095d3cc95260b3aab694ab06f179f83bcc flavour=1 folder=C:\ProgramData\Docker\tmp\hcs925889499

> docker inspect 6962984ea314

[
   ...
        "LogPath": "",
       ...
            "LogConfig": {
                "Type": "none",
                "Config": {}
            },
```

This is an inspect from a currently running container:
``
[
    ...
        "LogPath": "C:\\ProgramData\\Docker\\containers\\8d86b9609fdb611ee9a096178dd0667c9c8d79cec6689b0516c270e7782e277c\\8d86b9609fdb611ee9a096178dd0667c9c8d79cec6689b0516c270e7782e277c-json.log",
       ...
            "LogConfig": {
                "Type": "json-file",
                "Config": {}
            },
``

> docker version
Client:
 Version:       17.12.0-ce
 API version:   1.35
 Go version:    go1.9.2
 Git commit:    c97c6d6
 Built: Wed Dec 27 20:05:22 2017
 OS/Arch:       windows/amd64

Server:
 Engine:
  Version:      17.12.0-ce
  API version:  1.35 (minimum version 1.24)
  Go version:   go1.9.2
  Git commit:   c97c6d6
  Built:        Wed Dec 27 20:15:52 2017
  OS/Arch:      windows/amd64
  Experimental: true
``

**Comment by sandersaares** [comment]
I have started getting "hcsshim::ImportLayer failed in Win32: The system cannot find the file specified." after changing how I install containers during a Server 2016 VM build. After investigation it became apparent that some 1709 images were getting pulled due to configuration error. After correcting the scripts to not reference 1709 images, the error went away.

**Comment by pdefreitas** [comment]
The same here, even when I am using winpty as the cmd as someone recommended in docker/for-win.

**Comment by wmadrid** [comment]
@sandersaares I am getting a similar error "hcsshim::ImportLayer failed in Win32: Access is denied. (0x5) after I run a yarn commando over a microsoft/nanoserver:10.0.14393.2007 , the image has all folders and path variables to run nodejs over windows nanoserver.
Just after executing "RUN yarn" gets until "Done in x sec" and then the error. I tried to look for places where 1709 could be downloaded but I wasn't successful. 
docker: 17.06.2-ee-4, build dd2c358 over windows server 2016

**Comment by fhilton** [comment]
I was getting the "ImportLayer failed in Win32"  error when trying to build a mssql-server-windows dockerfile. I solved the error by increasing the container size and then rebuilding with --no-cache: https://akosdudas.wordpress.com/2017/12/29/docker-image-build-with-error-importlayer-failed-in-win32-the-system-cannot-find-the-path-specified/

**Comment by jhowardmsft** [comment]
Close this issue as it's external to moby at this point. Several docker bugs have been fixed.  The outstanding issue is the bugcheck of the utility VM itself during shutdown. There is a Windows fix for RS1 and RS3 images slated (which does not mean guaranteed) for release in the April round of Windows updates, aka '2018/4B'. Internal references 16117946 and 16117937. Images which are built from an updated base should no longer exhibit the issue.

**Comment by ColinDabritzViewpoint** [comment]
This appears to have happened to me because of default storage limits, see: https://stackoverflow.com/questions/46539165/docker-windows-build-fails-with-error-the-system-cannot-find-the-path-specifie

Upping the default container base size addressed the issue.

**Comment by yusuf-gunaydin** [comment]
@jhowardmsft Is there any chance to get these updates early? (Contacting Microsoft maybe?) We have been working around the bugs related to this one for months. Now we are stuck and cannot go any further without this being fixed.

**Comment by jhowardmsft** [comment]
@yusuf-gunaydin Not that I know of. The servicing bugs are still being triaged (it looks like earliest is May round of update for release) and AFAIK, there's not even a private fix been built yet.

**Comment by yusuf-gunaydin** [comment]
Thanks.

I'm not sure if it is a permanent solution, but using Windows 10 Insider Preview 17120.1 (rs4_release) and microsoft/windowsservercore-insider I could overcome https://github.com/docker/for-win/issues/1052. (I have tried a few times and all of it worked without any error.)

However adding `--squash` to the build command still produces the same error.

**Comment by jclay** [comment]
For anyone else who ends up here from Google like me, I was trying to create the build tools container as described in the official [Microsoft Documentation](https://docs.microsoft.com/en-us/visualstudio/install/build-tools-container#step-5-create-and-build-the-dockerfile) (followed verbatim). I was encountering this error every time I tried to build the image, but the following steps has my build working consistently:

1) Disable "Container" and "Hyper-V" from Windows Features.
2) Uninstall Docker
3) Reboot
4) Re-enable "Container" and "Hyper-V". Restart.
5) Download and install latest Docker.
6) Switch to using the 1709 tag of the windows server container.

```diff
- FROM microsoft/windowsservercore:1709
+ FROM microsoft/windowsservercore:latest
```

I'm pretty confident that switching to 1709 did not solve the problem, as I had applied this before the reinstall steps and it did not resolve. It could have been some combination of cached images, etc that were causing this to fail which gets resolved by a reinstall of Docker.

**Comment by sandersaares** [comment]
To add another useful hint for layer-related issues like this, the docker-ci-zap tool is just awesome in cleaning up the local cache of Docker image layers that are almost impossible to delete manually: https://github.com/jhowardmsft/docker-ci-zap

**Comment by jorisdg** [comment]
I did a mini-version of @jclay that worked for me.
My (large) windows containers started failing after the large windows 10 update in May. Couldn't rebuild the images due to this error. I uninstalled DfW, disabled containers, re-enabled containers, reinstalled DfW. No reboots anywhere in between. Everything is now working again.

**Comment by DNikitenko** [comment]
For me on Windows 10 installing 1803 update resolved the issue.

**Comment by jhowardmsft** [comment]
Thought I'd close the loop on this as the fix was released as part of the June 2018 patch Tuesday updates (aka 6B). Below is an edited snippet from internal email I just sent, which explains how to resolve. If anyone is hitting this using one of the updated images and latest daemon, it's now a different cause, so please open a new issue.


Oh, and ping @tianon :heart: :smile:. Sorry, forgot to tell you offline.
----

First, make sure you are running the latest docker daemon. There were various fixes, some of which are linked to in this issue above.

The next most likely cause is https://github.com/moby/moby/issues/32838#issuecomment-369399688. The underlying issue for that is the utility VM bug-checking during shutdown.

Both the RS1 and RS3 backports referred to in my comment above  were available in 2018/5C as a hotfix, which means they should have been part of the patch Tuesday update in 2018/6B (most recent one as at the time of writing, 6/22/18).

Note this is a fix to the utility VM image, not the host. It only affects Hyper-V containers (which means if you are running on Client SKUs, or if explicit and using Windows Server 2016 as a host), and only if you are running an image based from a downlevel (assuming you are running RS4 on the host, that means RS1 and RS3) image. It does not affect RS4 images on RS4 hosts, but IIRC, it does affect RS3 on RS3, RS1 on RS3 and RS1 on RS1.

If you update your Dockerfiles FROM statement so that it picks up an image based on 2018-6B, it shouldn’t happen anymore. So that would be 1709_KB4284819 for RS3 and 10.0.14393.2312 for RS1, based on tags at https://hub.docker.com/r/microsoft/windowsservercore/tags/.  

For example: 

- FROM microsoft/windowsservercore:1709_KB4284819 for an RS3 base image, OR

- FROM microsoft/windowsservercore:latest  (assuming you delete your current microsoft/windowsservercore image), or FROM microsoft/windowsservercore:10.0.14393.2312 for an RS1 base image.

Phew 😓

**Comment by thaJeztah** [comment]
Thanks for the update @jhowardmsft 

Let me lock this issue, so that that last comment does not get lost; as mentioned above, if you still run into this situation, please open a new ticket with details; thanks!
